---
title: "Final Project"
author: Caleb Maddry
date: March 22, 2005
output:
    md_document:
      variant: markdown_strict+backtick_code_blocks+autolink_bare_uris
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(GenomicRanges)
library(IRanges)
library(dplyr)
library(tidyr)
library(tibble)
library(readr)
library(ggplot2)
library(purrr)
library(magrittr)
library(pheatmap)
library(textshape)
library(Rcpp)
library(DESeq2)
library(stringr)
```

##### (a) Loading in ATACseq peak files with custom function import_peaks (list of GRanges output)

```{r loading in peak files to list of GRanges}

# establishing peak path to the dir with MACS2 output peak files from NF_CORE ATACseq pipeline
#peak_path <- "<your_atac_pipeline_output_dir>/bwa/merged_library/macs2/broad_peak"
peak_path <- "/scratch/Shares/rinnclass/MASTER_CLASS/lessons/08_ATACseq_pipeline/00_run_pipeline/00_pipeline_run/pipeline_run/bwa/merged_library/macs2/broad_peak"

# creating a file list also needed for import_peaks function to get sample name associated with file
fl <- list.files(peak_path, full.names = TRUE, pattern = ".broadPeak")


print("here are the number of peaks for each sample")

```
